rebalance/rebalance.py

```diff
@@ -161,23 +165,66 @@ def rebalance(plugin, outgoing_scid, incoming_scid, msatoshi: Millisatoshi = Non
     route_in = {'id': my_node_id, 'channel': incoming_scid, 'direction': int(not incoming_node_id < my_node_id)}
     start_ts = int(time.time())
     label = "Rebalance-" + str(uuid.uuid4())
     description = "%s to %s" % (outgoing_scid, incoming_scid)
     invoice = plugin.rpc.invoice(msatoshi, label, description, retry_for + 60)
     payment_hash = invoice['payment_hash']
-    success_msg = ""
-    try:
-        excludes = []
-        # excude all own channels to prevent unwanted shortcuts [out,mid,in]
-        mychannels = plugin.rpc.listchannels(source=my_node_id)['channels']
-        for channel in mychannels:
-            excludes += [channel['short_channel_id'] + '/0', channel['short_channel_id'] + '/1']
 
+    rpc_result = None
+    excludes = [my_node_id]  # excude all own channels to prevent shortcuts
+    nodes = {}               # here we store erring node counts
+    routes = {}              # trace all routes to suppress duplicates
+    maxhops = 1              # start with short routes and increase
+    msatfactor = plugin.msatfactor # start with high msatoshi factor to reduce
+                             # WIRE_TEMPORARY failures because of imbalances
+
+    # 'disable' maxhops filter if set to <= 0
+    if plugin.maxhops <= 0:
+        maxhops = 20
+
+    # trace stats
+    count = 0
+    count_sendpay = 0
+    time_getroute = 0
+    time_sendpay = 0
+
+    try:
         while int(time.time()) - start_ts < retry_for and not plugin.rebalance_stop:
-            r = plugin.rpc.getroute(incoming_node_id, msatoshi, riskfactor=1, cltv=9, fromid=outgoing_node_id, exclude=excludes)
+            count += 1
+            try:
+                time_start = time.time()
+                r = plugin.rpc.getroute(incoming_node_id,
+                                        fromid=outgoing_node_id,
+                                        exclude=excludes,
+                                        msatoshi=msatoshi * msatfactor,
+                                        maxhops=maxhops,
+                                        riskfactor=10, cltv=9)
+                time_getroute += time.time() - time_start
+            except RpcError as e:
+                # could not find route -> change params and restart loop
+                if e.method == "getroute" and e.error.get('code') == 205:
+                    # reduce msatfactor to look for smaller channels now
+                    msatfactor -= 1
+                    if msatfactor == 0:
+                        # when we reached neutral msat factor:
+                        # increase maxhops and restart with msatfactor
+                        maxhops = maxhops + 1
+                        msatfactor = plugin.msatfactor
+                        continue
+                    # we observed that we only have a ~3% success rate on routes that
+                    # are 8 hops or longer. Hence we cut at 6 (+2 in/out).
+                    if plugin.maxhops > 0 and maxhops > plugin.maxhops:
```

Comment:
Don't `continue` in the `if msatfactor == 0:` case. `maxhops` will increased here, so we should check `if plugin.maxhops > 0 and maxhops > plugin.maxhops:`

Reply:
By the way, you should really remove this `continue`. With `rebalance-msatfactor: 1` we never reach the maxhops check.

And the `if msatfactor == 0:` could be `if msatfactor < 1:`, as msatfactor is float.

Reply:
Thanks. Fixed by 0e5bf03
I moved the maxhops `if plugin.maxhops > 0 and maxhops > plugin.maxhops`in front of the `msatfactor` check. Sounds correct, right?

Reply:
Also please don't mind my stupid brain on  8f5f325  :D
I will sqash in the end ^^